ui: create physical query plan inspector admin view

jrbotros opened the pull request:
Provides a physical query plan inspector at /queryplan in the admin
interface, moving @RaduBerinde's cockroachdb.github.io code into the core
codebase. Committed with love for my roommate @benesch, since he's been
professing the Cockroach bible to me pretty much daily.

Address #12651.

(Also, there's definitely a security flaw here with dropping the user-provided query directly into an `EXPLAIN` statement—happy to write something more robust with direction from you!)

<img width="1374" alt="screen shot 2017-05-14 at 7 32 39 pm" src="https://cloud.githubusercontent.com/assets/2992765/26038700/21fc61be-38dc-11e7-9fc1-6f42aff54a0f.png">

jrbotros committed:
ui: create physical query plan inspector admin view

Provides a physical query plan inspector at /queryplan in the admin
interface, moving @RaduBerinde's cockroachdb.github.io code into the core
codebase. Committed with love for my roommate @benesch, since he's been
professing the Cockroach bible to me pretty much daily.

Address #12651.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15908)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=15908) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=15908) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=15908) it.</sub>

RaduBerinde commented:
Nice! I had no idea this was going on. I just made some recent changes to the viewer though:
https://github.com/cockroachdb/cockroachdb.github.io/pull/3
https://github.com/cockroachdb/cockroachdb.github.io/pull/4

Ideally we would share the common code somehow.

These plans are only for distsql queries, not general query plans as the address and title suggest. Also, we should display if the query would actually be run through DistSQL in auto mode (this is indicated by a column returned by `EXPLAIN (DISTSQL)`).

bdarnell commented:
Review status: 0 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 911 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCAJ1PmRNIquYT7pzc:-KkCAJ1PmRNIquYT7pzd:b-btd9hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L911)):*
> ```Go
> 	ctx context.Context, req *serverpb.PhysicalQueryPlanRequest,
> ) (*serverpb.PhysicalQueryPlanResponse, error) {
> 	args := sql.SessionArgs{User: s.getUser(req)}
> ```

`s.getUser` is currently just a stub that hard-codes "root" ; this will allow anyone on the network to see whatever metadata is exposed through this interface (or worse, see my next comment). We need proper authentication for the UI before we can expose more complex SQL-level functionality. 

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkC9kW2s1Niw7isNheF:b-dpx773) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
> ```Go
> 	defer session.Finish(s.server.sqlExecutor)
> 	explain := fmt.Sprintf(
> 		"SELECT JSON FROM [EXPLAIN (distsql) %s]",
> ```

This is vulnerable to SQL injection. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks for working on this, pretty cool!

We can look later at generating the `js` for the static website from the same `tsx` file (as suggested by @mrtracy), but it doesn't have to be done in this PR.

---

Review status: 0 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkD1Y11lApndnJ7IH0W:bruhgb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is vulnerable to SQL injection. 
</blockquote></details>

One way to avoid this would be to support passing a query as a string to `EXPLAIN`. The executor would parse the string and wrap it into an `explainNode`, and we would pass the string here as a placeholder.  Maybe @knz has other ideas.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-:-KkD24gWdwpp-xrUorE8:b3cb5u0)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
In my opinion, the common code should be the typescript code being created with this PR.  The sharable display code will need to be moved into a distinct package, and we can create a TS project that compiles that code into the JS code needed by the github.io page.  We'll have to figure out what that structure looks like in terms of our codebase, but that would be the goal.

For now, once the PR feedback here is addressed (mostly around the injection opportunity, but also a number of typescript issues to address) I believe we can get this checked into our code base and then begin moving towards shared code.  That would including porting Radu's latest changes to typescript, separating the shareable code into a module, and using that module both in our Admin UI *and* a new typescript project targeting the github.io viewer.

Thanks for the contribution!

---

Review status: 0 of 18 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD-YkXYlqOMDm1fQCd:-KkD-YkXYlqOMDm1fQCe:b-mox7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L908)):*
> ```Go
> }
> 
> func (s *adminServer) PhysicalQueryPlan(
> ```

Add a brief comment explaining what this does, that will bring it in line with the other methods in this file. 

---

*[pkg/server/admin.go, line 911 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCAJ1PmRNIquYT7pzc:-KkCKWuF-Rzr8MagoKGk:buk1rz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L911)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`s.getUser` is currently just a stub that hard-codes "root" ; this will allow anyone on the network to see whatever metadata is exposed through this interface (or worse, see my next comment). We need proper authentication for the UI before we can expose more complex SQL-level functionality. 
</blockquote></details>

We already allow anyone with access to the Admin UI to see the entire schema of all databases, and to see metrics information about all nodes on the cluster. Is this exposing information that is any more sensitive?

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkCLQUJe0ZMZs8rBkcA:b-uq7fm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is vulnerable to SQL injection. 
</blockquote></details>

To be clear, the string being explained should be provided to the query as a parameter, not placed directly in the string.  See the method "SetUIData" in this file for an example of doing this.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD2tM0jGlK8rqn6Lra:-KkD2tM1HCKvZ4-AeQSI:buscp7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L12)):*
> ```XML
> 
> interface DataGraph {
>   nodes: any[];
> ```

These should not be `any` collections, you should define the interface for each of these fields.  This is not coming in over the wire, you are creating these in the functions here.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 18 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD-lamCUNkAPWd8ohE:-KkD-lan0b0TaZE-7dJl:b-m6c817) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L18)):*
> ```XML
> }
> 
> export class QueryPlanGraph extends React.Component<QueryPlanGraphProps, {}> {
> ```

Because it is exported, write a short comment for this class.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCWWyO2xMM0W4Q92oY:-KkCWWyO2xMM0W4Q92oZ:b6u52h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L74)):*
> ```XML
> 
>   redrawPlan(jsonString: string) {
>     this.vis.selectAll("*").remove();
> ```

I'm assuming that this d3 code is just copied from Radu's original implementation?  It seems to make sense, I'm not going to review it in depth for now since it's still under active development. 

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 79 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD35joJGuPjCTSUwNv:-KkD35joJGuPjCTSUwNw:bhygjsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L79)):*
> ```XML
>     }
> 
>     const data = JSON.parse(jsonString);
> ```

This should be in a try-catch block, we can't assume that the passed JSON will parse correctly, and if an exception gets thrown inside of `render()` it's fairly catastrophic for react.  At this point, the component should likely just render a message like "invalid JSON" or something to that effect.

You should also test if the returned object is undefined, or if it's missing fields you think should be present.  Typescript gives us compile-time safety, but it does not give us runtime verification that a parsed JSON actually has all expected fields.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCWnUC08Ti2Aquj2xn:-KkCWnUDzeSKI58YGYZv:bqq8sct) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L89)):*
> ```XML
>       .attr("height", Math.min(containerWidth, containerHeight));
> 
>     this.groupsLayer = this.vis.append("g");
> ```

Why do these need to be attached to the class?  As far as I can tell, they are not being accessed outside of this function and should just be `const` variables.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 102 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCXFJZZm-60Us3M9VE:-KkCXFJZZm-60Us3M9VF:b-hai1mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L102)):*
> ```XML
>     graph.nodes = [];
>     for (let i = 0; i < data.processors.length; i++) {
>       let p = data.processors[i];
> ```

I think that a number of these `let` statements should be `const`.  A few of these variables do get mutated, but the majority are definitely not mutated.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD7Y-zoQCvBnQsm5ox:-KkD7Y-zoQCvBnQsm5oy:bsmybtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L104)):*
> ```XML
>       let p = data.processors[i];
>       p.core.graphNodeIdx = graph.nodes.length;
>       graph.nodes.push({
> ```

This is where typescript typechecking shines; if `graph.nodes` was a typed collection instead of `any[]`, the compiler would find discrepancies in any of the multiple places you push objects into the collection.  As it is, there is no type checking going on here and we lose confidence that the code is correct.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 229 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCXLQ8FHp6Ww4aX_fj:-KkCXLQ8FHp6Ww4aX_fk:btqml1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L229)):*
> ```XML
>     }
> 
>     let color = d3.scale.category20();
> ```

Should definitely be a `const`.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCXNipBC8Nr_v4ey0e:-KkCXNipBC8Nr_v4ey0f:bxdtzhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L252)):*
> ```XML
>         .attr("fill", "#000");
> 
>     let group = this.groupsLayer.selectAll(".group")
> ```

`const`.  A lot of these should be `const`, I won't bother annotating all of them.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 273 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD746L6_FVkcjI2OkY:-KkD746L6_FVkcjI2OkZ:b-nnkebq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L273)):*
> ```XML
>       .data(graph.nodes)
>       .enter().append("rect")
>       .attr("class", (d: any) => d.type)
> ```

If you type the DataGraph collections correctly, you won't need to attach these explicit `any` types to these functions.

Implicit `any` is disallowed; explicit `any` is a code smell and should require a strong justification whenever it is rarely used.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCUbzJ4ObMiBg2Poq1:-KkCUbzJ4ObMiBg2Poq2:b30u5op) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L6)):*
> ```XML
> import { connect } from "react-redux";
> import CodeMirror from "react-codemirror";
> import "codemirror/mode/sql/sql";
> ```

Being a side-effect import, this is a bit odd.  Can you add a comment that it has a side-effect of making SQL highlighting info available to CodeMirror?

---

*[pkg/ui/src/containers/queryPlan.tsx, line 8 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCUGGZeGDUzw8rH0WE:-KkCUGGZeGDUzw8rH0WF:b7pi57k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L8)):*
> ```XML
> import "codemirror/mode/sql/sql";
> import "codemirror/lib/codemirror.css";
> import "codemirror/theme/neat.css";
> ```

Can you move these CSS imports to `index.tsx` with the other CSS imports?  I know it's only being used by components in this one file, but CSS has a terrible global namespace and I would like to explicitly enumerate all of the third-party CSS we are including in one place.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCVUG9fEBGn3onOJ75:-KkCVUGAyUR2lpaXl2IC:b1kkwew) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L94)):*
> ```XML
> 
>   render() {
>     let options = {
> ```

`const`

---

*[pkg/ui/src/containers/queryPlan.tsx, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCT1qTpmxDMHBYgj3w:-KkCT1qTpmxDMHBYgj3x:buezc4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L104)):*
> ```XML
>         <CodeMirror
>           value={this.state.query}
>           onChange={this.updateCode.bind(this)}
> ```

Instead of this, define `updateCode` above as:

```typescript
updateCode = (query: string) => {
  // ...
}
```

That's an es6-ism supported by typescript that automatically binds the method to the instance, meaning you don't have to use the `bind` method.  It is a better implementation because using `bind` inside of `render()` creates a new function every time `render` is called.

Note that after doing this you may have to move `updateCode` a bit to please our linter, which enforces an ordering of declarations within classes.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCTheyH6o4vcnMl5tP:-KkCTheyH6o4vcnMl5tQ:b-ynvzws) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L110)):*
> ```XML
>       <div className="query-plan__graph">
>         {this.getQueryError()}
>         <button className="query-plan__refresh" onClick={this.getQueryPlan.bind(this)}>
> ```

Same as above, declare `getQueryPlan` as a bound function.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCS_csmpCE1QDXB2sN:-KkCS_csmpCE1QDXB2sO:b-dklp2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L122)):*
> ```XML
> let queryPlanConnected = connect(
>   (state: AdminUIState) => {
>     const cached = state.cachedData.physicalQueryPlan as any;
> ```

Why does this need to be cast to "any"?  You aren't doing anything exotic here and this is not coming from a third party component.

---

*[pkg/ui/styl/components/queryPlanGraph.styl, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCRXgZSu518z7mLmIw:-KkCRXgZSu518z7mLmIx:b-8h9jdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/styl/components/queryPlanGraph.styl#L1)):*
> ```Stylus
> .query-plan__graph
> ```

Normally I would suggest that you apply BEM-styling to all of these nodes instead of placing them under a parent like this (part of the motivation for BEM is to minimize cascading).  I also don't care for the non-specificity of many of these classes (link, label, etc.), this puts us at risk of getting unintended styling.

However, I also understand that this is derived from code written by Radu and that there will be a merger process, so I'm okay with this for now.  However, could you add a comment to this file about fixing this once we merge on one code base?

Also, if you *really* wanted extra credit, I would move this to `shame.styl`, which is our catch-all file for CSS that we are currently accepting but want to change when we get a chance to refactor.

---

*[pkg/ui/styl/layout/layout.styl, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCQtVkO2YMhP9A6EfJ:-KkCQtVkO2YMhP9A6EfK:bi6b7yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/styl/layout/layout.styl#L27)):*
> ```Stylus
>   display: flex
>   flex-direction: column
>   height: 100vh;
> ```

I'm generally okay with this change, but this does affect content on every other page, not just the one you're working on.  Would it be possible to not include this as part of this commit?

---

*[pkg/ui/styl/pages/queryplan.styl, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCQf8D7QU9fx1_nMJU:-KkCQf8D7QU9fx1_nMJV:b-33oyl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/styl/pages/queryplan.styl#L26)):*
> ```Stylus
>   position: relative;
> 
>   .query-plan__refresh
> ```

You've correctly assigned BEM classes here, but these should not be children of `query-plan__graph`; just move them up to the top level.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-:-KkCXccfHJ8eNsHAPz1F:bow3kz4)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 18 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkD8KPxEGLSVv6NgX_p:b-p08fo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

To be clear, the string being explained should be provided to the query as a parameter, not placed directly in the string.  See the method "SetUIData" in this file for an example of doing this.
</blockquote></details>

Ah, if EXPLAIN doesn't support placeholders then that is indeed an issue, and this would be blocked on that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 18 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkDo8vERglLUZmUaZuY:b-8vuxt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Ah, if EXPLAIN doesn't support placeholders then that is indeed an issue, and this would be blocked on that.
</blockquote></details>

I think another way to avoid issues here would be to expose an `ExecuteStatement` that only works if the sql parses to a single statement. There is no danger of injection in this particular case: it either doesn't parse, or it parses as an explain statement. But this is a bit fragile (it relies on the particularities of `EXPLAIN` which can't really be combined with other statements).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908)*
<!-- Sent from Reviewable.io -->

jrbotros commented:
Would it be worth publishing a `react-distsql-query-plan` npm package exporting this QueryPlanGraph component? Then we could just add and use dependency in both the admin code and `cockroachdb.github.io` (we'd have to set up a build system for the latter, but that's probably not a bad idea in any case).

Also, since the endpoint will ultimately return multiple query plans, @benesch advised me to change PhysicalQueryPlan to QueryPlan everywhere, so I've done that in this iteration.

Happy to help!

---

Review status: 0 of 17 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[pkg/server/admin.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD-YkXYlqOMDm1fQCd:-KkERUZ0uL2s_075XdY8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L908)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Add a brief comment explaining what this does, that will bring it in line with the other methods in this file. 
</blockquote></details>

Done.

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkERglIsnrRjv3hL_ld:bryj7x0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think another way to avoid issues here would be to expose an `ExecuteStatement` that only works if the sql parses to a single statement. There is no danger of injection in this particular case: it either doesn't parse, or it parses as an explain statement. But this is a bit fragile (it relies on the particularities of `EXPLAIN` which can't really be combined with other statements).
</blockquote></details>

Took a stab at this with some guidance from @benesch, let me know if it does the job!

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD2tM0jGlK8rqn6Lra:-KkERzdV85pGHGHWLdfC:bp5tdai) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L12)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

These should not be `any` collections, you should define the interface for each of these fields.  This is not coming in over the wire, you are creating these in the functions here.
</blockquote></details>

For sure—was having some issues with WebCola types and the original query plan code. Submitted tgdwyer/WebCola#228, we'll need to wait for that to be merged and published to npm before landing this (or we can make a fork with these changes, but the author seems pretty responsive so it's probably reasonable to wait it out).

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 18 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD-lamCUNkAPWd8ohE:-KkETe1IYvJo7jhnMbJy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L18)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Because it is exported, write a short comment for this class.
</blockquote></details>

Done.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCWWyO2xMM0W4Q92oY:-KkETfBaMdPtts6Y4x-c:bffanxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L74)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I'm assuming that this d3 code is just copied from Radu's original implementation?  It seems to make sense, I'm not going to review it in depth for now since it's still under active development. 
</blockquote></details>

Yeah, converted it to Typescript but otherwise kept the implementation close to the original.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 79 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD35joJGuPjCTSUwNv:-KkETx0uWxmpYqSQ1tWP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L79)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This should be in a try-catch block, we can't assume that the passed JSON will parse correctly, and if an exception gets thrown inside of `render()` it's fairly catastrophic for react.  At this point, the component should likely just render a message like "invalid JSON" or something to that effect.

You should also test if the returned object is undefined, or if it's missing fields you think should be present.  Typescript gives us compile-time safety, but it does not give us runtime verification that a parsed JSON actually has all expected fields.


</blockquote></details>

Done.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCWnUC08Ti2Aquj2xn:-KkETzA0ziCR7NqYBWQv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L89)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Why do these need to be attached to the class?  As far as I can tell, they are not being accessed outside of this function and should just be `const` variables.
</blockquote></details>

Done.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 102 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCXFJZZm-60Us3M9VE:-KkETzjznS8kM0HAOyio:bg1tdx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L102)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I think that a number of these `let` statements should be `const`.  A few of these variables do get mutated, but the majority are definitely not mutated.
</blockquote></details>

Fixed and enforced with #15941!

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD7Y-zoQCvBnQsm5ox:-KkEU4V3QLUJjNWtzHIo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L104)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This is where typescript typechecking shines; if `graph.nodes` was a typed collection instead of `any[]`, the compiler would find discrepancies in any of the multiple places you push objects into the collection.  As it is, there is no type checking going on here and we lose confidence that the code is correct.
</blockquote></details>

Done.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 229 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCXLQ8FHp6Ww4aX_fj:-KkEUC7DwjBAGLlF2K7g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L229)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Should definitely be a `const`.
</blockquote></details>

Done.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCXNipBC8Nr_v4ey0e:-KkEUDm9pYBNn5myUODB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L252)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

`const`.  A lot of these should be `const`, I won't bother annotating all of them.
</blockquote></details>

Done.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 273 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkD746L6_FVkcjI2OkY:-KkEUFTCbNjo9fKlLBiG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L273)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

If you type the DataGraph collections correctly, you won't need to attach these explicit `any` types to these functions.

Implicit `any` is disallowed; explicit `any` is a code smell and should require a strong justification whenever it is rarely used.
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCUbzJ4ObMiBg2Poq1:-KkEUNgY0_arOMUdxF3c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L6)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Being a side-effect import, this is a bit odd.  Can you add a comment that it has a side-effect of making SQL highlighting info available to CodeMirror?
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 8 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCUGGZeGDUzw8rH0WE:-KkEUOuwTqXLOTjIedyd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L8)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Can you move these CSS imports to `index.tsx` with the other CSS imports?  I know it's only being used by components in this one file, but CSS has a terrible global namespace and I would like to explicitly enumerate all of the third-party CSS we are including in one place.
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCVUG9fEBGn3onOJ75:-KkEUQ2oV-vU2mYTf1-H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L94)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

`const`
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCT1qTpmxDMHBYgj3w:-KkEURuG2YfvaL5lVP5x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L104)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Instead of this, define `updateCode` above as:

```typescript
updateCode = (query: string) => {
  // ...
}
```

That's an es6-ism supported by typescript that automatically binds the method to the instance, meaning you don't have to use the `bind` method.  It is a better implementation because using `bind` inside of `render()` creates a new function every time `render` is called.

Note that after doing this you may have to move `updateCode` a bit to please our linter, which enforces an ordering of declarations within classes.
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCTheyH6o4vcnMl5tP:-KkEUSdA5CsNbyXz-X8d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L110)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Same as above, declare `getQueryPlan` as a bound function.
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/queryPlan.tsx, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCS_csmpCE1QDXB2sN:-KkEUTzsRMfGJc_bV5Tu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/containers/queryPlan.tsx#L122)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Why does this need to be cast to "any"?  You aren't doing anything exotic here and this is not coming from a third party component.
</blockquote></details>

Done.

---

*[pkg/ui/styl/pages/queryplan.styl, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCQf8D7QU9fx1_nMJU:-KkEV2kUnGd3WUVgCzPe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/styl/pages/queryplan.styl#L26)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

You've correctly assigned BEM classes here, but these should not be children of `query-plan__graph`; just move them up to the top level.
</blockquote></details>

Done.

---

*[pkg/ui/styl/components/queryPlanGraph.styl, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCRXgZSu518z7mLmIw:-KkEZy2QLyhBraJ8-QXP:b13dgh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/styl/components/queryPlanGraph.styl#L1)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Normally I would suggest that you apply BEM-styling to all of these nodes instead of placing them under a parent like this (part of the motivation for BEM is to minimize cascading).  I also don't care for the non-specificity of many of these classes (link, label, etc.), this puts us at risk of getting unintended styling.

However, I also understand that this is derived from code written by Radu and that there will be a merger process, so I'm okay with this for now.  However, could you add a comment to this file about fixing this once we merge on one code base?

Also, if you *really* wanted extra credit, I would move this to `shame.styl`, which is our catch-all file for CSS that we are currently accepting but want to change when we get a chance to refactor.
</blockquote></details>

Since classes are derived from the `type` attr added to graph nodes, I think it's probably worth a separate refactor to clean this up. Added to `shame.styl`!

---

*[pkg/ui/styl/layout/layout.styl, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCQtVkO2YMhP9A6EfJ:-KkEUoAiJ7D3guVRgvwJ:b-mkj0fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/styl/layout/layout.styl#L27)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I'm generally okay with this change, but this does affect content on every other page, not just the one you're working on.  Would it be possible to not include this as part of this commit?
</blockquote></details>

Moved to #15945.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-:-KkEVBQKXxGah_5IZuCl:b-1rmq9o)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 17 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkG0QGCZF01hzJoUMlU:bdg199u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, jrbotros (Joseph Botros) wrote…</i></summary><blockquote>

Took a stab at this with some guidance from @benesch, let me know if it does the job!
</blockquote></details>

This looks good to me.. @bdarnell?

---

*[pkg/server/admin.go, line 929 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkG0LF3zHL4XsUJo4gO:-KkG0LF3zHL4XsUJo4gP:by9zyro) ([raw file](https://github.com/cockroachdb/cockroach/blob/1df15e4582d8b02edb0930f4e7b7aeaddd1deb75/pkg/server/admin.go#L929)):*
> ```Go
>   }
> 
>   log.Infof(ctx, req.Query)
> ```

Duplicate log.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 17 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkG4IWt1TgiDtcR_qwb:baeg2kp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This looks good to me.. @bdarnell?
</blockquote></details>

This looks good to me too. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908)*
<!-- Sent from Reviewable.io -->

jrbotros commented:
Review status: 0 of 17 files reviewed at latest revision, 23 unresolved discussions.

---

*[pkg/server/admin.go, line 929 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkG0LF3zHL4XsUJo4gO:-KkGKw4pukAFrOypUFN0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1df15e4582d8b02edb0930f4e7b7aeaddd1deb75/pkg/server/admin.go#L929)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Duplicate log.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
I think this PR is good, module a few comments and waiting on the upstream changes to WebCola typings.

In regards to an NPM module, I think we'd like to avoid the overhead of publishing a module; it might not be much, but it's conceptually disconcerting to use a third party service to link our own modules together.

@tamird has suggested using git submodules, which is what we do for vendored code in Go; we would add the distsql-plan-graph code to a new repo under the CockroachDB organization, then submodule that repo into both the main Cockroach project *and* a new project which builds the github.io page.

---

Review status: 0 of 17 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 102 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCXFJZZm-60Us3M9VE:-KkH7WVYo6XKyPSqltbT:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/ui/src/components/queryPlanGraph.tsx#L102)):*
<details><summary><i>Previously, jrbotros (Joseph Botros) wrote…</i></summary><blockquote>

Fixed and enforced with #15941!
</blockquote></details>

Nice.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 59 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkH67QbzZW1GZwYh6C_:-KkH67QbzZW1GZwYh6Ca:b-omc2uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fad9106f8b807b221eb7cd079a38912ecffed43/pkg/ui/src/components/queryPlanGraph.tsx#L59)):*
> ```XML
>   vis: d3.Selection<SVGElement>;
>   zoomRect: d3.Selection<SVGElement>;
>   groupsLayer: d3.Selection<SVGElement>;
> ```

Some of these are no longer being used, I think only the first three are being used anymore.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 64 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkH6kZlvLbfqkMJ8qI0:-KkH6kZlvLbfqkMJ8qI1:b-6pdgph) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fad9106f8b807b221eb7cd079a38912ecffed43/pkg/ui/src/components/queryPlanGraph.tsx#L64)):*
> ```XML
> 
>   // parseError is set if invalid JSON is provided in props.
>   parseError = false;
> ```

I don't see this being used anywhere, there's no need to assign to it if you aren't going to display a message.  This should either be removed, or it should be the basis of displaying some sort of error message to the user.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-:-KkH9rD2QcTT0eUwm1_c:b-icnnfi)*
<!-- Sent from Reviewable.io -->

jrbotros commented:
The submodule approach seems reasonable to me! WebCola typings are all set and I've made the other changes, now just waiting on the TeamCity build.

---

Review status: 0 of 17 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 59 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkH67QbzZW1GZwYh6C_:-KkIodlk58GS60lCozJ6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fad9106f8b807b221eb7cd079a38912ecffed43/pkg/ui/src/components/queryPlanGraph.tsx#L59)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Some of these are no longer being used, I think only the first three are being used anymore.
</blockquote></details>

Done.

---

*[pkg/ui/src/components/queryPlanGraph.tsx, line 64 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkH6kZlvLbfqkMJ8qI0:-KkIoe702QkZzkWnpgQh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fad9106f8b807b221eb7cd079a38912ecffed43/pkg/ui/src/components/queryPlanGraph.tsx#L64)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I don't see this being used anywhere, there's no need to assign to it if you aren't going to display a message.  This should either be removed, or it should be the basis of displaying some sort of error message to the user.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-:-KkIoik6jivtPZYmb2Ow:bktzit3)*
<!-- Sent from Reviewable.io -->

benesch commented:
TeamCity failed due to what looks like a difference in the way the latest version of Yarn generates its lockfile. Agh! Regenerated the yarn.lock file—hopefully that does the trick. Heads up, @jrbotros—I force-pushed your branch with the fix.

RaduBerinde commented:
Hats off to how quickly and effectively you addressed the comments! Thanks for the contribution!

LGTM if @mrtracy is happy with the typescript stuff.

mrtracy commented:
LGTM, I am rebasing and force-pushing to resolve the conflicts with generated files

bdarnell commented:
Review status: 0 of 17 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 911 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkCAJ1PmRNIquYT7pzc:-KkMRoegd-dDMq4iVYDq:b-76fbvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L911)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

We already allow anyone with access to the Admin UI to see the entire schema of all databases, and to see metrics information about all nodes on the cluster. Is this exposing information that is any more sensitive?
</blockquote></details>

No, this shouldn't expose any more data than we already do with the schema (at least for now - when we have stats-based query optimization then query plans might leak statistical information that is not otherwise available), so it's OK for now. But we should fix this - at the SQL level we don't let get schema metadata for tables you don't have any permissions on, so we should enforce a similar rule in the UI. 

---

*[pkg/server/admin.go, line 915 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908#-KkC9kW2s1Niw7isNheE:-KkMSsg1hrhowVM5BH7y:bb6ywmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d946ee847b13285566f815d841cc4c9a8b6effe3/pkg/server/admin.go#L915)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This looks good to me too. 
</blockquote></details>

LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15908)*
<!-- Sent from Reviewable.io -->
